File: loxide/src/syntax.rs
//! Loxide syntax trees

use std::fmt;

use crate::{
    srcloc::SrcLoc,
    value::Value,
};

/// A unary operator
#[derive(Copy, Clone, Debug)]
pub enum UnOp {
    Complement,
    Negate,
}

impl fmt::Display for UnOp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            UnOp::Complement => write!(f, "!"),
            UnOp::Negate => write!(f, "-"),
        }
    }
}

/// A binary operator
#[derive(Copy, Clone, Debug)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    NotEq,
    Lt,
    LtEq,
    Gt,
    GtEq,
    And,
    Or,
}

impl fmt::Display for BinOp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            BinOp::Add => write!(f, "+"),
            BinOp::Sub => write!(f, "-"),
            BinOp::Mul => write!(f, "*"),
            BinOp::Div => write!(f, "/"),
            BinOp::Eq => write!(f, "=="),
            BinOp::NotEq => write!(f, "!="),
            BinOp::Lt => write!(f, "<"),
            BinOp::LtEq => write!(f, "<="),
            BinOp::Gt => write!(f, ">"),
            BinOp::GtEq => write!(f, ">="),
            BinOp::And => write!(f, "and"),
            BinOp::Or => write!(f, "or"),
        }
    }
}

/// A Lox expression
#[derive(Clone, Debug)]
pub enum Expr<V> {
    Literal(Value, SrcLoc),
    UnOpApp(UnOp, Box<Expr<V>>, SrcLoc),
    BinOpApp(BinOp, Box<Expr<V>>, Box<Expr<V>>, SrcLoc),
    Var(V, SrcLoc),
    Assign(V, Box<Expr<V>>, SrcLoc),
    Call(Box<Expr<V>>, Vec<Expr<V>>, SrcLoc),
    This(V, SrcLoc),
    PropertyGet(Box<Expr<V>>, String, SrcLoc),
    PropertySet(Box<Expr<V>>, String, Box<Expr<V>>, SrcLoc),
    /* this is hacky: super has two variable slots
     *   one for the superclass
     *   one for "this"
     */
    Super(V, V, String, SrcLoc),
}

impl<V> Expr<V> {
    pub fn location(&self) -> &SrcLoc {
        match self {
            Expr::Literal(_, loc) => loc,
            Expr::UnOpApp(_, _, loc) => loc,
            Expr::BinOpApp(_, _, _, loc) => loc,
            Expr::Var(_, loc) => loc,
            Expr::Assign(_, _, loc) => loc,
            Expr::Call(_, _, loc) => loc,
            Expr::This(_, loc) => loc,
            Expr::PropertyGet(_, _, loc) => loc,
            Expr::PropertySet(_, _, _, loc) => loc,
            Expr::Super(_, _, _, loc) => loc,
        }
    }
}

impl<V: fmt::Display> fmt::Display for Expr<V> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Expr::Literal(v, _) => write!(f, "{}", v),
            Expr::UnOpApp(op, e, _) => write!(f, "{}{}", op, e),
            Expr::BinOpApp(op, lhs, rhs, _) =>
              write!(f, "({} {} {})", lhs, op, rhs),
            Expr::Var(v, _) => write!(f, "{}", v),
            Expr::Assign(v, e, _) => write!(f, "({} = {})", v, e),
            Expr::Call(callee, args, _) => {
                write!(f, "({})(", callee)?;
                let mut sep = "";
                for a in args {
                    write!(f, "{}{}", sep, a)?;
                    sep = ", "
                }
                write!(f, ")")
            },
            Expr::This(_, _) => write!(f, "this"),
            Expr::PropertyGet(o, name, _) => write!(f, "{}.{}", o, name),
            Expr::PropertySet(o, name, e, _) =>
              write!(f, "({}.{} = {})", o, name, e),
            Expr::Super(_, _, name, _) => write!(f, "super.{}", name),
        }
    }
}

/// A function or method definition
#[derive(Clone, Debug)]
pub struct FunOrMethod<V, S> {
    pub parameters: Vec<V>,
    pub body: Vec<Stmt<V, S>>,
    pub slots: S,
    pub location: SrcLoc,
}

/// A Lox statement
#[derive(Clone, Debug)]
pub enum Stmt<V, S> {
    Expr(Expr<V>, SrcLoc),
    IfElse(Expr<V>, Box<Stmt<V, S>>, Option<Box<Stmt<V, S>>>, SrcLoc),
    While(Expr<V>, Box<Stmt<V, S>>, SrcLoc),
    Print(Expr<V>, SrcLoc),
    Return(Option<Expr<V>>, SrcLoc),
    Block(Vec<Stmt<V, S>>, S, SrcLoc),
    VarDecl(V, Option<Expr<V>>, SrcLoc),
    FunDef(V, FunOrMethod<V, S>),
    /* this is hacky: the superclass has two "variable slots":
     *   the one pointing to the superclass (this is "read")
     *   the one where the superclass object gets stored (this is "written")
     */
    ClassDef(V, Option<(V, V)>, Vec<(String, FunOrMethod<V, S>)>, SrcLoc),
}

impl<V, S> Stmt<V, S> {
    pub fn location(&self) -> &SrcLoc {
        match self {
            Stmt::Expr(_, loc) => loc,
            Stmt::IfElse(_, _, _, loc) => loc,
            Stmt::While(_, _, loc) => loc,
            Stmt::Print(_, loc) => loc,
            Stmt::Return(_, loc) => loc,
            Stmt::Block(_, _, loc) => loc,
            Stmt::VarDecl(_, _, loc) => loc,
            Stmt::FunDef(_, f) => &f.location,
            Stmt::ClassDef(_, _, _, loc) => loc,
        }
    }
}

impl<V: fmt::Display, S> fmt::Display for Stmt<V, S> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Stmt::Expr(e, _) => write!(f, "{};", e),
            Stmt::IfElse(cond, s_if, s_else, _) => {
                write!(f, "if ({}) {{\n{}\n}}", cond, s_if)?;
                if let Some(s) = s_else {
                    write!(f, " else {{\n{}\n}}", s)?;
                }
                Ok(())
            },
            Stmt::While(cond, body, _) =>
                write!(f, "while ({}) {{\n{}\n}}", cond, body),
            Stmt::Print(e, _) => write!(f, "print {};", e),
            Stmt::Return(Some(e), _) => write!(f, "return {};", e),
            Stmt::Return(None, _) => write!(f, "return;"),
            Stmt::Block(body, _, _) => {
                write!(f, "{{\n")?;
                for s in body {
                    write!(f, "{}\n", s)?;
                }
                write!(f, "}}")
            },
            Stmt::VarDecl(v, init, _) => {
                write!(f, "var {}", v)?;
                if let Some(e) = init {
                    write!(f, " = {}", e)?;
                }
                write!(f, ";")
            },
            Stmt::FunDef(name, def) => {
                write!(f, "fun {} (", name)?;
                let mut sep = "";
                for p in &def.parameters {
                    write!(f, "{}{}", sep, p)?;
                    sep = ", ";
                }
                write!(f, ") {{\n")?;
                for s in &def.body {
                    write!(f, "{}\n", s)?;
                }
                write!(f, "}}")
            },
            Stmt::ClassDef(name, sup, methods, _) => {
                write!(f, "class {}", name)?;
                if let Some((cls, _)) = sup {
                    write!(f, " < {}", cls)?;
                }
                write!(f, " {{\n")?;
                for (name, m) in methods {
                    write!(f, "{} (", name)?;
                    let mut sep = "";
                    for p in &m.parameters {
                        write!(f, "{}{}", sep, p)?;
                        sep = ", ";
                    }
                    write!(f, ") {{\n")?;
                    for s in &m.body {
                        write!(f, "{}\n", s)?;
                    }
                    write!(f, "}}\n")?;
                }
                write!(f, "}}")
            },
        }
    }
}

File: saumon_fumé/lib/syntax.ml
type unary_op =
  | Complement
  | Negate

let pprint_unary_op = function
  | Complement -> "!"
  | Negate -> "-"

type binary_op =
  | Add
  | Sub
  | Mul
  | Div
  | Eq
  | NotEq
  | Lt
  | LtEq
  | Gt
  | GtEq
  | And
  | Or

let pprint_binary_op = function
  | Add -> "+"
  | Sub -> "-"
  | Mul -> "*"
  | Div -> "/"
  | Eq -> "=="
  | NotEq -> "!="
  | Lt -> "<"
  | LtEq -> "<="
  | Gt -> ">"
  | GtEq -> ">="
  | And -> "and"
  | Or -> "or"

module type Spec = sig
  type var
  val pprint_var: var -> string

  type scope_info

  type 'a annot
  val pprint_annot: ('a -> string) -> 'a annot -> string
end

module type S = sig
  type var
  type scope_info
  type 'a annot

  val pprint_var: var annot -> string

  type expr =
    | Lit of Value.t
    | UnaryOp of unary_op * expr annot
    | BinaryOp of binary_op * expr annot * expr annot
    | Var of var annot
    | Assign of var annot * expr annot
    | Call of expr annot * expr annot list

  val pprint_expr: expr annot -> string

  type stmt =
    | Expr of expr annot
    | IfElse of expr annot * stmt annot * stmt annot option 
    | While of expr annot * stmt annot
    | Print of expr annot
    | Return of expr annot option
    | Block of stmt annot list * scope_info
    | VarDecl of var annot * expr annot option
    | FunDef of var annot * var annot list * stmt annot list * scope_info

  val pprint_stmt: stmt annot -> string

  type prog = stmt annot list

  val pprint_prog: prog -> string
end

module Make (Sp: Spec): S
with type var = Sp.var
and type scope_info = Sp.scope_info
and type 'a annot = 'a Sp.annot = struct
  type var = Sp.var
  type 'a annot = 'a Sp.annot

  type scope_info = Sp.scope_info

  let pprint_var = Sp.pprint_annot Sp.pprint_var

  type expr =
    | Lit of Value.t
    | UnaryOp of unary_op * expr annot
    | BinaryOp of binary_op * expr annot * expr annot
    | Var of var annot
    | Assign of var annot * expr annot
    | Call of expr annot * expr annot list

  let rec pprint_expr' = function
    | Lit v -> Value.pprint v
    | UnaryOp(op, e) ->
        let e' = pprint_expr e in
        Printf.sprintf "%s%s" (pprint_unary_op op) e'
    | BinaryOp(op, lhs, rhs) ->
        let lhs' = pprint_expr lhs in
        let rhs' = pprint_expr rhs in
        Printf.sprintf "(%s %s %s)" lhs' (pprint_binary_op op) rhs'
    | Var v -> pprint_var v
    | Assign (v, e) ->
        let v' = pprint_var v in
        let e' = pprint_expr e in
        Printf.sprintf "(%s = %s)" v' e'
    | Call (callee, args) ->
        let callee' = pprint_expr callee in
        let args' = String.concat ", " (List.map pprint_expr args) in
        Printf.sprintf "(%s)(%s)" callee' args'
  and pprint_expr e = Sp.pprint_annot pprint_expr' e

  type stmt =
    | Expr of expr annot
    | IfElse of expr annot * stmt annot * stmt annot option 
    | While of expr annot * stmt annot
    | Print of expr annot
    | Return of expr annot option
    | Block of stmt annot list * scope_info
    | VarDecl of var annot * expr annot option
    | FunDef of var annot * var annot list * stmt annot list * scope_info

  let rec pprint_stmt' = function
    | Expr e -> pprint_expr e ^ ";"
    | IfElse (e, s1, s2) ->
        let els = match s2 with
          | None -> ""
          | Some s -> " else " ^ pprint_stmt s
        in
        "if (" ^ pprint_expr e ^ ") " ^ pprint_stmt s1 ^ els
    | While (e, s) ->
        "while (" ^ pprint_expr e ^ ") " ^ pprint_stmt s
    | Print e -> "print " ^ pprint_expr e ^ ";"
    | Return None -> "return;"
    | Return (Some e) -> "return " ^ pprint_expr e ^ ";"
    | Block (stmts, _) ->
        "{\n" ^ String.concat "\n" (List.map pprint_stmt stmts) ^ "\n}"
    | VarDecl (v, None) -> "var " ^ pprint_var v ^ ";"
    | VarDecl (v, Some e) -> "var " ^ pprint_var v ^ " = " ^ pprint_expr e ^ ";"
    | FunDef (name, params, body, _) ->
        let name' = pprint_var name in
        let params' = String.concat ", " (List.map pprint_var params) in
        let body' = String.concat "\n" (List.map pprint_stmt body) in
        Printf.sprintf "%s (%s) {\n%s\n}" name' params' body'
  and pprint_stmt s = Sp.pprint_annot pprint_stmt' s

  type prog = stmt annot list

  let pprint_prog p =
    String.concat "" (List.map (fun s -> pprint_stmt s ^ "\n") p)
end

module AsParsed = Make (struct
  type var = string
  let pprint_var v = v

  type scope_info = unit

  type 'a annot = 'a Located.t
  let pprint_annot = Located.pprint
end)

module AsResolved = Make (struct
  type var = string * int * int
  let pprint_var (name, _, _) = name

  type scope_info = int

  type 'a annot = 'a Located.t
  let pprint_annot = Located.pprint
end)
